### DISCUSSION TITLE
How to use UnoCss locally and create custom preset

### DISCUSSION THREAD

**Discussion by ghost** [unocss/unocss#3013] [Q&A] [answered]
Hi, I'trying to use UnoCSS locally. Here is my progress:

I download the core ( without any preset ) from the CDN ( `https://cdn.jsdelivr.net/npm/@unocss/runtime/core.global.js` ) into my project folder.

The struture looks like this:
apple
--index.php
--unocss.js ( the one download from the CDN )
--custom.js
--config.js

In the `custom.js` ( I followed the instruction here: https://unocss.dev/config/presets ):
```
import { Preset } from '/unocss.js';

export default function myPreset() {
    return {
        name: 'my-preset',
        rules: [
            [/^text-(\w+)$/, ([, c]) => `color: ${c}`]
        ]
    }
}
```
1) The browser throws an error: `Uncaught SyntaxError: The requested module '/unocss.js' does not provide an export named 'Preset'`.
2) My second question is, how the `config.js` looks like? Is it something like this:
```
import { defineConfig } from '/unocss.js';
import myPreset from '/custom.js';

export default defineConfig({
    presets: [
        myPreset(),
    ]
})
```
3) What do we need to give a name `name: 'my-preset'` in `custom.js` because we specify which preset we want to use by the function name like `presets: [ myPreset({ /* preset options */ }) }` in `config.js` instead of `presets: [ my-preset ]`?

**Comment by userquin** [comment]
Remove the import in your preset, it is just VanillaJS and not being used.

**Comment by ghost** [comment]
@userquin It still not work:
```
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
    <script src="unocss.js"></script>
    <script src="custom.js"></script>
    <script src="config.js"></script>
</head>

<body>
    <p class="text-blue">Hello</p>
</body>

</html>
```

`custom.js`
```
export default function myPreset() {
    return {
        name: 'my-preset',
        rules: [
            [/^text-(\w+)$/, ([, c]) => `color: ${c}`]
        ]
    }
}
```

`config.js`
```
export default defineConfig({
    presets: [
        myPreset(),
    ],
})
```

**Comment by chz** [comment]
Here how can you use runtime unocss inside html:
https://jsbin.com/xaruqameve/edit?html,js,output

**Comment by ghost** [comment]
@chz Thank you for your reference! Unfortunately I still don't see the word becoming blue :(

https://jsfiddle.net/0ymhk2v9/

**Reply by userquin** [reply]
Change the UnoCSS runtime script to: `<script src="https://cdn.jsdelivr.net/npm/@unocss/runtime"></script>`

![imagen](https://github.com/unocss/unocss/assets/6311119/83ab66b6-8452-4d2a-8776-9e1c135c89c8)

**Reply by userquin** [reply]
https://unocss.dev/integrations/runtime

**Comment by ghost** [comment]
@userquin Hi, the reason I use `https://cdn.jsdelivr.net/npm/@unocss/runtime/core.global.js` is because I want to create my own preset, I don't want to load the default preset.

**Reply by userquin** [reply, marked as answer]
so the problem is using `fiddlejs`, if you add the script at top the unocss runtime will run before the configuration script, remove the script tag and change your script to this one:
```js
window.__unocss = {
  rules: [
     [/^text-(\w+)$/, ([, c]) => {
       console.log('CALLING RULE')
       return `color: ${c}`
     }]
  ], 
  presets: [],
  // ...
}

import('https://cdn.jsdelivr.net/npm/@unocss/runtime/core.global.js')
```

![imagen](https://github.com/unocss/unocss/assets/6311119/234097b5-b950-4acc-b181-dd7d73e79f70)

**Reply by userquin** [reply]
I also tried to reload the configuration via `window.__unocss_runtime.update()`, but it seems it is cached.

**Reply by userquin** [reply]
you can also add `type="module"` to the script tag, it will be loaded with defer:
![imagen](https://github.com/unocss/unocss/assets/6311119/89352a1e-07e9-4ad0-9856-c3aea7823e07)

**Comment by ghost** [comment]
@userquin Thank you very much for your honest help.

1) Is it possible to create a custom preset with runtime?
2) I saw the doc and and read your code, if we can set empty for preset and add rules with `window._unocss`, actually what is the reason of creating a custom preset? Any differences between them? Just curious.
